core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -167,13 +175,96 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
         )
     }
 
-  private def getSnapshotFromDataRepo(snapshotIdentifiers: NamedDataRepoSnapshot) =
-    Try(dataRepoDAO.getSnapshot(snapshotIdentifiers.snapshotId, ctx.userInfo.accessToken)) match {
+  // Finds a workspace using the workspaceId then calls the createSnapshot method
+  def createSnapshotsByWorkspaceIdV3(workspaceId: String, snapshotIds: Set[UUID]): Future[Unit] =
+    getV2WorkspaceContextAndPermissionsById(workspaceId,
+                                            SamWorkspaceActions.write,
+                                            Some(WorkspaceAttributeSpecs(all = false))
+    ).flatMap { rawlsWorkspace =>
+      createSnapshots(rawlsWorkspace, snapshotIds)
+    }
+
+  // Find a workspace using the workspaceName then calls the createSnapshot method
+  def createSnapshotsByWorkspaceNameV3(workspaceName: WorkspaceName, snapshotIds: Set[UUID]): Future[Unit] =
+    getV2WorkspaceContextAndPermissions(workspaceName,
+                                        SamWorkspaceActions.write,
+                                        Some(WorkspaceAttributeSpecs(all = false))
+    ).flatMap(rawlsWorkspace => createSnapshots(rawlsWorkspace, snapshotIds))
+
+  // Link the snapshot pao to the workspace pao
+  private def createSnapshots(rawlsWorkspace: Workspace, snapshotIds: Set[UUID]): Future[Unit] =
+    for {
+      snapshotsFromDataRepo <- Future {
+        snapshotIds.map(getSnapshotFromDataRepoWithId)
+      }
+      workspacePaoOpt <- policyService.getPao(rawlsWorkspace.workspaceIdAsUUID, ctx)
+      workspacePao = workspacePaoOpt.getOrElse(
+        throw new RawlsExceptionWithErrorReport(
+          ErrorReport(StatusCodes.NotFound,
+                      s"Workspace PAO not found for workspace ${rawlsWorkspace.workspaceIdAsUUID}"
+          )
+        )
+      )
+
+      snapshotPaos <- Future.traverse(snapshotsFromDataRepo) { snapshot =>
+        policyService.getOrCreateSnapshotPao(snapshot.getId, ctx)
+      }
+
+      _ <- Future.traverse(snapshotsFromDataRepo) { snapshot =>
```

Comment:
I am not convinced that this provides much value. This will check that each individual snapshot can link to the workspace, but it will not guarantee that there will not be conflicts by combining the snapshot PAOs together. Would love thoughts on this.